### ISSUE TITLE
Speed issue when adding linear processing pathway

### ISSUE THREAD

**Issue by tylergiallanza** [PrincetonUniversity/PsyNeuLink#1955]
It appears that the `add_linear_processing_pathway` method for Compositions is inefficient for large mechanisms. Consider the following code sample:

```
comp = pnl.Composition()
mech1 = pnl.TransferMechanism(default_variable=np.zeros(100))
mech2 = pnl.TransferMechanism(default_variable=np.zeros(1000))
comp.add_linear_processing_pathway([mech1,mech2])
```

This takes 32.3 seconds to run on my machine (using the `%%timeit` magic in Jupyter notebook). Creating the pathway manually, however:
```
comp = pnl.Composition()
mech1 = pnl.TransferMechanism(default_variable=np.zeros(100))
mech2 = pnl.TransferMechanism(default_variable=np.zeros(1000))
comp.add_node(mech1)
comp.add_node(mech2)
map_ = pnl.MappingProjection(matrix=np.random.rand(100, 1000),
                               name="map_h12"
                            )
comp.add_projection(projection=map_,sender=mech1,receiver=mech2)
```
Takes only 3 seconds to run.

P.S. the time to run the `add_linear_processing_pathway` seems to scale super-linearly w.r.t the size of `mech2` in the above example:
| # of Units in Mech2  | Time to create composition |
| ------------- | ------------- |
| 100  | 1 second  |
| 1000  | 30 seconds  |
| 2000  | 90 seconds  |

**Comment by jdcpni** [comment]
FE(everyone’s)I:  its in the call to analyze_graph (25 of the 30 seconds).   More to explore.


jdc



On Mar 18, 2021, at 5:22 PM, tylergiallanza ***@***.******@***.***>> wrote:



It appears that the add_linear_processing_pathway method for Compositions is inefficient for large mechanisms. Consider the following code sample:

comp = pnl.Composition()
mech1 = pnl.TransferMechanism(default_variable=np.zeros(100))
mech2 = pnl.TransferMechanism(default_variable=np.zeros(1000))
comp.add_linear_processing_pathway([mech1,mech2])


This takes 32.3 seconds to run on my machine (using the %%timeit magic in Jupyter notebook). Creating the pathway manually, however:

comp = pnl.Composition()
mech1 = pnl.TransferMechanism(default_variable=np.zeros(100))
mech2 = pnl.TransferMechanism(default_variable=np.zeros(1000))
comp.add_node(mech1)
comp.add_node(mech2)
map_ = pnl.MappingProjection(matrix=np.random.rand(100, 1000),
                               name="map_h12"
                            )
comp.add_projection(projection=map_,sender=mech1,receiver=mech2)


Takes only 3 seconds to run.

P.S. the time to run the add_linear_processing_pathway seems to scale super-linearly w.r.t the size of mech2 in the above example:

# of Units in Mech2     Time to create composition
100     1 second
1000    30 seconds
2000    90 seconds

**Comment by jdcpni** [comment]
Tracked down what is most of the problem (that scales with array size): inefficiency in validating arrays within utilities.is_compatible.  Posted a patch to devel that cuts the time down by a factor of 3;  should be on master with the next pull.  The rest has to do with a call to _analyze_graph(), which is made automatically by the add_linear_processing_pathway, but not (currently) by the more specific add methods. However, the same price will be paid at run time with those methods.  In any event, with the correction just made, it should no longer scale with matrix/array sizes.


jdc



On Mar 18, 2021, at 8:19 PM, Jonathan D. Cohen ***@***.******@***.***>> wrote:

FE(everyone’s)I:  its in the call to analyze_graph (25 of the 30 seconds).   More to explore.

jdc



On Mar 18, 2021, at 5:22 PM, tylergiallanza ***@***.******@***.***>> wrote:



It appears that the add_linear_processing_pathway method for Compositions is inefficient for large mechanisms. Consider the following code sample:

comp = pnl.Composition()
mech1 = pnl.TransferMechanism(default_variable=np.zeros(100))
mech2 = pnl.TransferMechanism(default_variable=np.zeros(1000))
comp.add_linear_processing_pathway([mech1,mech2])


This takes 32.3 seconds to run on my machine (using the %%timeit magic in Jupyter notebook). Creating the pathway manually, however:

comp = pnl.Composition()
mech1 = pnl.TransferMechanism(default_variable=np.zeros(100))
mech2 = pnl.TransferMechanism(default_variable=np.zeros(1000))
comp.add_node(mech1)
comp.add_node(mech2)
map_ = pnl.MappingProjection(matrix=np.random.rand(100, 1000),
                               name="map_h12"
                            )
comp.add_projection(projection=map_,sender=mech1,receiver=mech2)


Takes only 3 seconds to run.

P.S. the time to run the add_linear_processing_pathway seems to scale super-linearly w.r.t the size of mech2 in the above example:

# of Units in Mech2     Time to create composition
100     1 second
1000    30 seconds
2000    90 seconds

**Comment by kmantel** [comment]
#1963 brings down the example 1 time significantly (sub 1s on the 100/2000 case). However it's still about 5x as slow as example 2. I think the discrepancy before and after suggests there is more here that should be looked into (maybe one example doesn't do everything that's needed, for example), and it's not obvious to me what the difference would be.

**Comment by kmantel** [comment]
Since _analyze_graph would need to be called at Composition.run-time anyway, I consider this fixed by #1961 and #1963
